django/contrib/postgres/fields/array.py

```diff
@@ -307,6 +307,17 @@ def get_prep_lookup(self):
                 prepared_values.append(tuple(value))
         return prepared_values
 
+    def batch_process_rhs(self, compiler, connection, rhs=None):
+        # Cast the RHS items to the LHS, because there are cases in which
+        # Postgres cannot infer the right type. For instance
+        # postgres_tests.test_array.TestQuerying.test_in fails with
+        # "operator does not exist: integer[] = numeric[]", although it's
+        # possible to cast manually numeric[]::integer[].
+        sqls, sqls_params = super().batch_process_rhs(compiler, connection, rhs)
+        cast_ph = f"%s::{self.lhs.field.cast_db_type(connection)}"
+        sqls = tuple(cast_ph if sql == "%s" else sql for sql in sqls)
+        return sqls, sqls_params
```

Comment:
I feel like the proper solution here is the same as the one used for JSONField in https://github.com/apollo13/django/pull/3.

Make sure the lookup that needs rhs handling subclass `FieldGetDbPrepValueMixin` and have `ArrayField.get_db_prep_value` return a subclass of `list` that has registered adapter on the connection to add the proper casts.

Reply:
Sadly not exactly. This is a little bit more tricky :/ The error received is:
```
django.db.utils.ProgrammingError: operator does not exist: integer[] = smallint[]
LINE 1: ...postgres_tests_nullableintegerarraymodel"."field" IN ($1, $2...
                                                             ^
HINT:  No operator matches the given name and argument types. You might need to add explicit type casts.

----------------------------------------------------------------------
(0.000)
SELECT "postgres_tests_nullableintegerarraymodel"."id",
       "postgres_tests_nullableintegerarraymodel"."field",
       "postgres_tests_nullableintegerarraymodel"."field_nested",
       "postgres_tests_nullableintegerarraymodel"."order"
FROM "postgres_tests_nullableintegerarraymodel"
WHERE "postgres_tests_nullableintegerarraymodel"."field" IN ('{1}'::int2[], '{2}'::int2[]);

args=([1],
      [2]);
```

The issue here is that psycopg checks the size of the passed integers and checks where an integer would fit. We can force the proper type by using `Int4` for the data. That said we cannot easily change Django's IntegerField to `Int4`, or should we add a `adapt_ipaddressfield_value` alike there?

Reply:
Fixed it like this for now: https://github.com/apollo13/django/commit/5e111c19620d00fec45f365c6d8cc1cc7419a53d

Open to better suggestions.